### PR TITLE
add some support for rename on case insensitive local filesystems

### PR DESCRIPTION
**Pull request by icewind1991**
Add *some* logic for handling case insensitive filesystems when `localstorage.case_insensitive` is set.

This adds a bunch of overhead as `file_exists` now needs to read the directory content to perform the case sensitive compare.

With my local testing (case insensitive zfs), the `rename('foo', 'Foo')` call returns true but doesn't actually change the on-disk name, so an additional check is added that the name has actually changed. So the rename fails but at least doesn't cause a filecache desync.

### CODE REVIEW

**Comment by icewind1991** [comment]
Note that having proper support for case insensitive file systems is a significant amount of work

**Comment by Altahrim** [comment]
Tested on `stable27`  :+1:

File: lib/private/Files/Storage/Local.php [hunk 1]
```diff
@@ -372,6 +387,11 @@ public function rename($source, $target): bool {
 		}
 
 		if (@rename($this->getSourcePath($source), $this->getSourcePath($target))) {
+			if ($this->caseInsensitive) {
+				if (strtolower($target) === strtolower($source) && !$this->file_exists($target)) {
```
Comment (Altahrim) [inline lib/private/Files/Storage/Local.php:391]: I think we should care to multibyte strings:
```suggestion
				if (mb_strtolower($target) === mb_strtolower($source) && !$this->file_exists($target)) {
```
I renamed a file `É` to `é` and I had the issue.

File: lib/private/Files/Storage/Local.php [hunk 2]
```diff
@@ -74,6 +74,8 @@ class Local extends \OC\Files\Storage\Common {
 
 	protected bool $unlinkOnTruncate;
 
+	private $caseInsensitive = false;
```
Comment (juliushaertl) [inline lib/private/Files/Storage/Local.php:77]:
```suggestion
	private bool $caseInsensitive = false;
```

File: lib/private/Files/Storage/Local.php [hunk 3]
```diff
@@ -93,6 +95,7 @@ public function __construct($arguments) {
 		$this->config = \OC::$server->get(IConfig::class);
 		$this->mimeTypeDetector = \OC::$server->get(IMimeTypeDetector::class);
 		$this->defUMask = $this->config->getSystemValue('localstorage.umask', 0022);
+		$this->caseInsensitive = $this->config->getSystemValue('localstorage.caseInsensitive', false);
```
Comment (kesselb) [inline lib/private/Files/Storage/Local.php:98]: We are already using

localstorage.umask
localstorage.unlink_on_truncate
localstorage.allowsymlinks

and therefore I would prefer localstorage.case_insensitive to keep it a bit consistent ;)

File: lib/private/Files/Storage/Local.php [hunk 4]
```diff
@@ -74,6 +74,8 @@ class Local extends \OC\Files\Storage\Common {
 
 	protected bool $unlinkOnTruncate;
 
+	private bool $caseInsensitive = false;
```
Comment (Altahrim) [inline lib/private/Files/Storage/Local.php:77]: May I suggest to add it as `protected` in `Common` so we'll be able to use it for other storages too

**Review by Altahrim — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Pull request merged**